bluemira/equilibria/optimisation/problem/_position.py

```diff
@@ -94,33 +95,46 @@ def __init__(
         self.opt_parameters = opt_parameters
         self._constraints = [] if constraints is None else constraints
 
-    def optimise(self, **_) -> CoilsetOptimiserResult:
+    def optimise(self, x0: npt.NDArray | None = None, **_) -> CoilsetOptimiserResult:
         """
         Run the optimisation.
 
         Returns
         -------
         The result of the optimisation.
         """
-        # Get initial state and apply region mapping to coil positions.
-        initial_state, _ = self.read_coilset_state(self.coilset, self.scale)
-        initial_x, initial_z, initial_currents = np.array_split(initial_state, 3)
-        initial_mapped_positions = self.position_mapper.to_L(initial_x, initial_z)
+        self.update_magnetic_constraints(I_not_dI=True, fixed_coils=False)
+
+        if x0 is None:
+            # Get initial state and apply region mapping to coil positions.
+            cs_opt_state = self.coilset.get_optimisation_state(
+                self.position_mapper.interpolator_names, current_scale=self.scale
+            )
+            initial_mapped_positions = self.position_mapper.to_L(
+                cs_opt_state.xs, cs_opt_state.zs
+            )
+
+            len_mapped_pos = len(initial_mapped_positions)
+            x0 = np.concatenate((initial_mapped_positions, cs_opt_state.currents))
+
         eq_constraints, ineq_constraints = self._make_numerical_constraints()
         opt_result = optimise(
-            f_objective=self.objective,
-            x0=np.concatenate((initial_mapped_positions, initial_currents)),
+            f_objective=lambda vector: self.objective(vector, len_mapped_pos),
```

Comment:
This will fail if `x0` is provided because `len_mapped_pos` will not be defined

Reply:
Good catch!